fathom/src/core/semantics.rs

```diff
@@ -52,11 +59,58 @@ pub enum Value<'arena> {
     ConstLit(Const),
 }
 
+/// Initialization operation for lazy values.
+///
+/// We need to use a [defunctionalized] representation because Rust does not
+/// allow closures of type `dyn (Clone + FnOnce() -> Arc<Value>)`.
+///
+/// [defunctionalized]: https://en.wikipedia.org/wiki/Defunctionalization
+#[derive(Clone, Debug)]
+enum LazyInit<'arena> {
+    EvalTerm(SharedEnv<ArcValue<'arena>>, &'arena Term<'arena>),
+    ApplyElim(Arc<LazyValue<'arena>>, Arc<Elim<'arena>>),
```

Comment:
Why do we need `Arc`s for `LazyValue` and `Elim`? Is this to keep the size down?

Reply:
Both are necessary to prevent recursion/infinite size error